tikv-client/src/test/java/com/pingcap/tikv/predicates/TiKVScanAnalyzerTest.java

```diff
@@ -408,4 +429,40 @@ private int[] getIndexLens() {
       assertEquals(t.isCovering, isCovering);
     }
   }
+
+  @Test
+  public void buildRangeWithUnsignedLongPKTest() {
+    TiTableInfo table = createTableWithUnsignedLong(6, 5);
+    TiIndexInfo pkIndex = TiIndexInfo.generateFakePrimaryKeyIndex(table);
+
+    Expression eq1 =
+        greaterThan(ColumnRef.create("c1", table), Constant.create(0, IntegerType.BIGINT));
+    Expression eq2 =
+        lessThan(ColumnRef.create("c1", table), Constant.create(10, IntegerType.BIGINT));
+
+    List<Expression> exprs = ImmutableList.of(eq1, eq2);
+
+    ScanSpec result = TiKVScanAnalyzer.extractConditions(exprs, table, pkIndex);
+    List<IndexRange> irs =
+        expressionToIndexRanges(
+            result.getPointPredicates(), result.getRangePredicate(), table, pkIndex);
+
+    TiKVScanAnalyzer scanAnalyzer = new TiKVScanAnalyzer();
+
+    Map<Long, List<Coprocessor.KeyRange>> keyRanges =
+        scanAnalyzer.buildTableScanKeyRange(table, irs, null);
+
+    assertEquals(keyRanges.size(), 1);
+
+    Coprocessor.KeyRange keyRange = keyRanges.get(table.getId()).get(0);
+
+    assertEquals(
+        ByteString.copyFrom(
+            new byte[] {116, -128, 0, 0, 0, 0, 0, 0, 6, 95, 114, 0, 0, 0, 0, 0, 0, 0, 1}),
+        keyRange.getStart());
+    assertEquals(
+        ByteString.copyFrom(
+            new byte[] {116, -128, 0, 0, 0, 0, 0, 0, 6, 95, 115, 0, 0, 0, 0, 0, 0, 0, 0}),
+        keyRange.getEnd());
```

Comment:
I have a question about your test. Your predicate is 10>c1>0, but the range you get is not (0,10).
could you explain the test more clearly?

Reply:
@shiyuhang0 

This is mainly to fix `TiExpressionException: Cannot encode row key with non-long type` when create table with BITINT UNSIGNED primary key.

So the actual generation of startkey and endkey is here:

https://github.com/pingcap/tispark/blob/ba30d8c74cdd36edff26bdfb2a5d98c52f7491a4/tikv-client/src/main/java/com/pingcap/tikv/predicates/TiKVScanAnalyzer.java#L326

https://github.com/pingcap/tispark/blob/ba30d8c74cdd36edff26bdfb2a5d98c52f7491a4/tikv-client/src/main/java/com/pingcap/tikv/predicates/TiKVScanAnalyzer.java#L340

As discussed above, this fixes the issue
https://github.com/pingcap/tispark/pull/2291#discussion_r847199850

Reply:
I am sorry that I don't pay much attention to this pr because I've been so busy the last two weeks. I am free now.
As for your solution:
It can work because TiSpark doesn't push down the filter with BIGINT UNSIGNED primary key, So Spark will do the filter instead. It will lead to the wrong answer if Tispark pushdown this filters to TiKV later.
I am going to check if it should be pushdown, if it can't be pushdown your solution can be accepted. otherwise, you should return the current range (start_key & end_key) to fetch the correct data from tikv